crates/matrix-sdk-ui/src/timeline/polls.rs

```diff
@@ -0,0 +1,169 @@
+        content: UnstablePollResponseEventContent,
+    ) -> Self {
+        let mut clone = self.clone();
+        clone.response_events.push(ResponseEvent { sender, timestamp, content });
+        clone
+    }
+
+    pub(super) fn end(
+        &self,
+        sender: OwnedUserId,
+        timestamp: MilliSecondsSinceUnixEpoch,
+        content: UnstablePollEndEventContent,
+    ) -> Self {
+        let mut clone = self.clone();
+        clone.end_event = Some(EndEvent { sender, timestamp, content });
+        clone
+    }
+
+    pub fn results(&self) -> FfiPoll {
+        let responses = self
+            .response_events
+            .iter()
+            .map(|e| OriginalSyncUnstablePollResponseEvent {
+                // Concocting an OriginalSyncUnstablePollResponseEvent as I'm not able to get it from the handle_event() fn.
+                content: e.content.clone(),
+                event_id: EventId::new(<&ServerName>::try_from("example.com").unwrap()), // TODO Remove example.com
+                sender: e.sender.clone(),
+                origin_server_ts: e.timestamp,
+                unsigned: MessageLikeUnsigned::new(),
+            })
+            .collect::<Vec<OriginalSyncUnstablePollResponseEvent>>();
+
+        let results = compile_unstable_poll_results(
+            &self.start_event.content.poll_start,
+            responses.iter().map(|i| i).collect::<Vec<&OriginalSyncUnstablePollResponseEvent>>(),
+            None,
+        );
+
+        FfiPoll {
+            question: self.start_event.content.poll_start.question.text.clone(),
+            kind: match self.start_event.content.poll_start.kind {
+                ruma::events::poll::start::PollKind::Undisclosed => FfiPollKind::Undisclosed,
+                ruma::events::poll::start::PollKind::Disclosed => FfiPollKind::Disclosed,
+                _ => FfiPollKind::Undisclosed, // Safe default. Should we keep it?
+            },
+            max_selections: self.start_event.content.poll_start.max_selections.into(),
+            answers: self
+                .start_event
+                .content
+                .poll_start
+                .answers
+                .iter()
+                .map(|i| FfiPollAnswer { id: i.id.clone(), text: i.text.clone() })
+                .collect(),
+            votes: results
+                .iter()
+                .map(|i| (i.0.to_string(), i.1.into_iter().map(|i| i.to_string()).collect()))
+                .collect(),
+            end_time: if let Some(e) = self.end_event.clone() {
+                Some(e.timestamp.0.into())
+            } else {
+                None
+            },
+        }
+    }
+}
+
+impl TryFrom<&TimelineItemContent> for PollState {
+    type Error = ();
+
+    fn try_from(value: &TimelineItemContent) -> Result<PollState, Self::Error> {
+        match value {
+            TimelineItemContent::Poll(poll_state) => Ok(poll_state.clone()),
+            _ => Err(()),
+        }
+    }
+}
```

Comment:
What is this impl used for? It feels a bit weird.

Reply:
Yeah I agree, it's used only from test code, I'll move it in the test module